File: Union Find/Weighted Union/WeightedUnion.cpp
#include <bits/stdc++.h>
using namespace std;

class WeightedUnion{
    vector<int> p,sz;
    public:
        WeightedUnion(int N) {
            p = vector<int>(N);
            sz = vector<int>(N);
            for (int i=0;i<N;i++){
                p[i] = i;
                sz[i] = 1;
            }
        }

        int find(int i){
            while(i != p[i]){
                i = p[i];
            }
            return i;
        }

        void wunion(int i, int j){
            int ri = find(i);
            int rj = find(j);
            if (ri != rj)
                if (sz[ri] < sz[rj]){
                    p[ri] = rj;
                    sz[rj] += sz[ri];
                } else {
                    p[rj] = ri;
                    sz[ri] += sz[rj];
                }
        }      
};


int main() {
    int N,M;
    cin>>N;
    cin>>M;
    WeightedUnion wu = WeightedUnion(N);
    for (int i=0;i<M;i++){
        char c;
        cin>>c;
        if (c=='F'){
            int x;
            cin>>x;
            cout<<wu.find(x)<<endl;
        } else if (c=='U') {
            int x,y;
            cin>>x>>y;
            wu.wunion(x,y);
        }
    }
}

File: Union Find/Weighted Union/WeightedUnion.py
class WeightedUnion:
    
    def __init__(self, N):
        self.sz = [1 for _ in range(N)]
        self.p = [i for i in range(N)]
    
    def find(self, i):
        while(i != self.p[i]):
            i = self.p[i]
        return i
    
    def union(self,i,j):
        ri = self.find(i)
        rj = self.find(j)
        if (ri != rj):
            if (self.sz[ri] < self.sz[rj]):
                self.p[ri] = rj
                self.sz[rj] += self.sz[ri]
            else:
                self.p[rj] = ri
                self.sz[ri] += self.sz[rj]


N = int(input())
M = int(input())
lines = [input().split() for _ in range(M)]
WU = WeightedUnion(N)
for line in lines:
    if line[0]=='F':
        print(WU.find(int(line[1])))
    elif line[0]=='U':
        WU.union(int(line[1]),int(line[2]))
